text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+crate provides bindings for many APIs found on the web but accessing the
+functionality of parent classes is quite cumbersome.
+
+The web makes extensive use of class inheritance hierarchies, and in `web-sys`
+right now each class gets its own `struct` type with inherent methods. These
+types implement `AsRef` between one another for subclass relationships, but it's
+quite unergonomic to actually reference the functionality! For example:
+
+```rust
+let x: &Element = ...;
+let y: &Node = x.as_ref();
+y.append_child(...);
+```
+
+or...
+
+```rust
+let x: &Element = ...;
+<Element as AsRef<Node>>::as_ref(x)
+    .append_child(...);
+```
+
+It's be much nicer if we could support this in a more first-class fashion and
+make it more ergonomic!
+
+[RFC 3]: https://github.com/rustwasm/rfcs/pull/3
+
+> **Note**: While this RFC has the same motivation as [RFC 3] it's proposing an
+> alternative solution, specifically enabled by switching by `structural` by
+> default, which is discussed in [RFC 3] but is hopefully formally outlined
+> here.
+
+# Detailed Explanation
+[detailed-explanation]: #detailed-explanation
+
+This RFC proposes using the built-in `Deref` trait to model the class hierarchy
+found on the web in `web-sys`. This also proposes changes to `#[wasm_bindgen]`
+to make using `Deref` feasible for binding arbitrary JS apis (such as those on
+NPM) with `Deref` as well.
+
+For example, `web-sys` will contain:
+
+```rust
+impl Deref for Element {
+    type Target = Node;
+
+    fn deref(&self) -> &Node { /* ... */ }
+}
+```
+
+allowing us to write our example above as:
+
+```rust
+let x: &Element = ...;
+x.append_child(...); // implicit deref to `Node`!
+```
+
+All JS types in `web-sys` and in general have at most one superclass. Currently,
+however, the `#[wasm_bindgen]` attribute allows specifying multiple `extends`
+attributes to indicate superclasses:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(extends = Node, extends = Object)]
+    type Element;
+
+    // ...
+}
+```
+
+The `web-sys` API generator currently lists an `extends` for all superclasses,
+transitively. This is then used in the code generator to generate `AsRef`
+implementatiosn for `Element`.
+
+The code generation of `#[wasm_bindgen]` will be updated with the following
+rules:
+
+* If no `extends` attribute is present, defined types will implement
+  `Deref<Target=JsValue>`.
+* Otherwise, the *first* `extends` attribute is used to implement
+  `Deref<Target=ListedType>`.
+
+This means that `web-sys` may need to be updated to ensure that the immediate
+superclass is listed first in `extends`. Manual bindings will continue to work
+and will have the old `AsRef` implementations as well as a new `Deref`
+implementation.
+
+The `Deref` implementation will concretely be implemented as:
+
+```rust
+impl Deref for #imported_type {
+    type Target = #target_type;
+
+    #[inline]
+    fn deref(&self) -> &#target_type {
```

Comment:
As above, I think this should be `&Self::Target`

Reply:
([same as above](https://github.com/rustwasm/rfcs/pull/5#discussion_r223448924))